pkg/querier/querier.go

```diff
@@ -193,55 +193,76 @@ func (q *Querier) LabelNames(ctx context.Context, req *connect.Request[typesv1.L
 	}), nil
 }
 
-func filterSeriesByLabelNames(labelSets []*typesv1.Labels, labelNameMap map[string]struct{}) {
-	if len(labelNameMap) == 0 {
-		return
-	}
-
-	for _, ls := range labelSets {
-		labelCount := 0
-		for idx := range ls.Labels {
-			_, ok := labelNameMap[ls.Labels[idx].Name]
-			if ok {
-				ls.Labels[labelCount] = ls.Labels[idx]
-				labelCount++
-			}
-		}
-		ls.Labels = ls.Labels[:labelCount]
-	}
-}
-
 func (q *Querier) Series(ctx context.Context, req *connect.Request[querierv1.SeriesRequest]) (*connect.Response[querierv1.SeriesResponse], error) {
 	sp, ctx := opentracing.StartSpanFromContext(ctx, "Series")
 	defer func() {
 		sp.LogFields(
 			otlog.String("matchers", strings.Join(req.Msg.Matchers, ",")),
 			otlog.String("label_names", strings.Join(req.Msg.LabelNames, ",")),
+			otlog.Int64("start", req.Msg.Start),
+			otlog.Int64("end", req.Msg.End),
 		)
 		sp.Finish()
 	}()
 
-	// build up map of label names
-	labelNameMap := make(map[string]struct{}, len(req.Msg.LabelNames))
-	for _, labelName := range req.Msg.LabelNames {
-		labelNameMap[labelName] = struct{}{}
+	// Some clients may not be sending us timestamps. If start or end are 0, then
+	// mark this a legacy request. Legacy requests only query the ingesters.
+	legacyRequest := req.Msg.Start == 0 || req.Msg.End == 0
+
+	if q.storeGatewayQuerier == nil || legacyRequest {
+		responses, err := q.seriesFromIngesters(ctx, &ingestv1.SeriesRequest{
```

Comment:
The legacy behaviour is was query from ingesters - Head index only.

With this PR it changes to Head Index and all of the blocks on the ingesters. I fear that this will be fairly slow. So I think something similar should also be done in the ingester Serires. 

```suggestion
		responses, err := q.seriesFromIngesters(ctx, &ingestv1.SeriesRequest{
```